sys/cam/cam_periph.c

```diff
@@ -932,7 +934,7 @@ cam_periph_mapmem(union ccb *ccb, struct cam_periph_map_info *mapinfo,
 	for (i = 0; i < numbufs; i++) {
 
 		/* Save the user's data address. */
-		mapinfo->orig[i] = *data_ptrs[i];
+		mapinfo->orig[i] = __USER_CAP(*data_ptrs[i], lengths[i]);
```

Comment:
I'm not super familiar with this part of the code, however it seems that `data_ptrs[]` holds `uint8_t*` and `struct ccb_dev_match` (for example) contains various pointers. I believe that I will have to do some `uint64_t` or `__ptr64` annotation for these to work in the purecap kernel. Do you know which userspace tool uses these ioctls?

Reply:
I don't know, and it could be out-of-tree things like smartctl, etc.  The pass(4) driver used for SCSI pass through commands uses this interface.  camcontrol might be another one.  However, the goal here is not to let freebsd64 versions work, but to let purecap versions work.  Perhaps data_ptrs[] array needs to be an explicit array of capabilities instead.

Ah, yes.  I need to instead go splat a bunch of __kerncap annotations in `union ccb` and rework some of this to assume user capabilities instead.  If we need freebsd64 support we will have to do that separately.

Reply:
I've reworked `union ccb` to actually be purecap friendly which removed all the `__USER_CAP` uses in this file.